File: API/src/main/java/com/ltizzi/ecommerce/model/stock/StockMapper.java
package com.ltizzi.ecommerce.model.stock;

import com.ltizzi.ecommerce.model.product.ProductEntity;
import com.ltizzi.ecommerce.model.product.ProductMapper;
import com.ltizzi.ecommerce.model.product.ProductResponse;
import com.ltizzi.ecommerce.model.stockEntry.StockEntryMapper;
import com.ltizzi.ecommerce.model.stockEntry.StockEntryResponse;
import com.ltizzi.ecommerce.repository.ProductRepository;
import com.ltizzi.ecommerce.repository.StockRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Leonardo Terlizzi
 */

@Component
public class StockMapper {

    @Autowired
    private ProductMapper prodMapper;

    @Autowired
    private ProductRepository prodRepo;

    @Autowired
    private StockEntryMapper entryMapper;

    @Autowired
    private StockRepository stockRepo;

    public StockResponse toStockResponse(StockEntity stock) {
        StockResponse stockRes = new StockResponse();
        stockRes.setStock_id(stock.getStock_id());

        ProductResponse product = prodMapper.toProductResponse(stock.getProduct());
        stockRes.setProduct(product);
        stockRes.setCantidad(stock.getCantidad());
//        List<StockEntryResponse> entries = entryMapper.toArrayStockEntryResponse(stock.getEntries());
//        stockRes.setEntries(entries);
        return stockRes;
    }

    public StockEntity toStockEntity(StockRequest stockReq) {
        StockEntity stock = new StockEntity();
        if (stockReq.getStock_id() != null) {
            stock = stockRepo.findById(stockReq.getStock_id()).orElseThrow();
            assert stock != null;
            stock.setStock_id(stockReq.getStock_id());
        }

        stock.setCantidad(stockReq.getCantidad());
        stock.setEntries(entryMapper.toArrayStockEntryEntity(stockReq.getEntries()));
        ProductEntity product = prodRepo.findById(stockReq.getProduct().getId()).orElseThrow();
        stock.setProduct(product);
        return stock;
    }

    public List<StockResponse> toArrayStockResponse(List<StockEntity> stocks) {
        List<StockResponse> stocksRes = new ArrayList<>();
        stocks.forEach(stock -> {
            stocksRes.add(toStockResponse(stock));
        });
        return stocksRes;
    }
}

File: Server/src/dto/mappers/stock.mapper.ts
import { Stock } from "../../models/Stock";
import { StockEntry } from "../../models/StockEntry";
import { StockService } from "../../services/StockService";
import { StockRequest } from "../requests/stock.request";
import { StockResponse } from "../responses/stock.response";
import { ProductMapper } from "./product.mapper";
import { StockEntryMapper } from "./stockEntry.mapper";

const prodMapper = new ProductMapper();
const stockServ = new StockService();
const entryMapper = new StockEntryMapper();

export class StockMapper {
  toStockResponse(stock: Stock): StockResponse {
    const stockRes = {} as StockResponse;
    stockRes.stock_id = stock.stock_id;
    console.log("FROM MAPPER");
    console.log(stock);
    stockRes.product = prodMapper.toProductResponse(stock.product);
    stockRes.cantidad = stock.cantidad;
    return stockRes;
  }

  async toStockEntity(fromStock: StockResponse | StockRequest): Promise<Stock> {
    let stock = {} as Stock;
    if (fromStock.stock_id) {
      stock = (await stockServ.getStockById(fromStock.stock_id)) as Stock;
    }
    if (fromStock.hasOwnProperty("entries")) {
      stock.entries = (await entryMapper.toArrayEntryEntity(
        fromStock.entries
      )) as Array<StockEntry>;
    }
    stock.cantidad = fromStock.cantidad;
    stock.product = await prodMapper.toProductEntity(fromStock.product);
    return stock;
  }

  toArrayStockResponse(stocks: Array<Stock>): Array<StockResponse> {
    let stocksRes = [] as Array<StockResponse>;
    stocks.forEach((stock) => stocksRes.push(this.toStockResponse(stock)));
    return stocksRes;
  }
}
